accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,431 @@
+
+Consider the following where `ref\out\in` all map to a reference (`&`) by the runtime:
+```cs
+class MethodHolder
+{
+    public static void M1(in int i) { }
+    public static void M2(ref int i) { }
+    public static void M3(out int i) { i = 42; }
+}
+
+Type t1 = typeof(MethodHolder).GetMethod("M1").GetParameters()[0].ParameterType;
+Type t2 = typeof(MethodHolder).GetMethod("M2").GetParameters()[0].ParameterType;
+Type t3 = typeof(MethodHolder).GetMethod("M3").GetParameters()[0].ParameterType;
+
+Debug.Assert(t1 == t2 && t2 == t3); // True today; all are "int reference type" (int&)
+```
+
+Also consider [ILGenerator.EmitCalli()](https://docs.microsoft.com/dotnet/api/system.reflection.emit.ilgenerator.emitcalli#system-reflection-emit-ilgenerator-emitcalli(system-reflection-emit-opcode-system-runtime-interopservices-callingconvention-system-type-system-type())) as a reference implementation which does not have `ref\out\in` modifiers:
+```cs
+void EmitCalli(
+    System.Reflection.Emit.OpCode opcode,
+    System.Runtime.InteropServices.CallingConvention unmanagedCallConv,
+    Type? returnType,
+    Type[]? parameterTypes)
+```
+
+So custom modifiers, except for calling convensions, are not exposed in the runtime. They are not necessary for the key scenario of passing or invoking a function pointer and would also interfere with equality and castability. If function pointers expose all custom modifiers, such as `ref\out\in` and the C++/CLI `const`, then equality and castability will be based on exact matches to that which is considered overly restrictive for runtime use and may break C++/CLI usage.
+
+The only custom modifiers exposed by the runtime are the built-in classes representing calling conventions - these classes are named `System.Runtime.InteropServices.CallConv*`:
+- [CallConvCdecl](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvcdecl)
+- [CallConvStdCall](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvstdcall)
+- [CallConvThiscall](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvthiscall)
+- [CallConvFastcall](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvfastcall)
+- [CallConvSuppressGCTransition](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvsuppressgctransition)
+- [CallConvMemberFunction](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvmemberfunction)
+
+C# in particular just maps calling conventions specified in a function pointer to `System.Runtime.InteropServices.CallConv*` types as metadata. These CallConv* types must live in the same assembly as defines `System.Object` (i.e. `System.Private.Corelib.dll`).
+
```

Comment:
> This would require a ECMA-335 change but that would be minimally breaking.

I'm not sure I agree with this. This would have required a change for every metadata reader to accept that this byte would be compressed. It would have meant that any tool that read in metadata on its own would have been broken when reading v6 assemblies that contained these new function pointers.

I'm not arguing there wasn't a better way, I am merely pushing back on this comment as I don't think the case was made it would have been "minimally breaking".

Reply:
I just referenced ECMA-335 on this to ensure my criticism was founded in something rather than ignorance on details. I think this would have technically been an impactful breaking change but looking at the various implementations it might have been "okay". `calli` takes a token to a "stand-alone signature" - III.3.20. The "stand-alone signature" layout is defined in II.23.2.3 where the following is stated:

> The first byte of the Signature holds bits for HASTHIS, EXPLICITTHIS and calling 
convention – DEFAULT, VARARG, C, STDCALL, THISCALL, or FASTCALL. These are 
OR’d together

The assumption above being it is defined to be a "byte". This would imply to me that most parsers would simply read a byte and move on. However, the current implementation of the reader helpers for stand-along signatures do call `CorSigUncompressData()` so I guess we might have gotten away with it. Although it would have limited us to 32 different calling conventions and based on how we are now using base and modifying calling conventions, that might be a bit limiting.

Regardless of that, I appreciate the "off-topic note" as it did push me to look into this area.

Reply:
For reference, https://github.com/dotnet/csharplang/pull/3441 is the main issue where the design was discussed.

> encode new calling conventions using custom modifiers

Roslyn used this encoding for other things before. IIRC, it was a no-brainer to follow the established patterns. This was not first use of custom modifier attached to return parameter to encode things that have nothing to do with return parameter. 

For example, compile and run the following:
```
Console.WriteLine(typeof(MyStruct).GetMethod("set_MyProperty").ReturnParameter.GetRequiredCustomModifiers()[0]);

public struct MyStruct
{
    public int MyProperty { get; init; }
}
```
You will see that `IsExternalInit` custom modifier attached to the return parameter.

> Although it would have limited us to 32 different calling conventions and based on how we are now using base and modifying calling conventions, that might be a bit limiting.

Right, there are hundreds of different calling conventions out there.

Also, one of the design goals of the encoding was that it should be possible to add new calling convention variants without modifying language or Roslyn compiler for each addition.

Reply:
If we were to do a breaking change to the file format for this, I think it would be better to add a special calling convention that just says (on a high level): the next byte of the signature specifies a number N. After this, there's N tokens. That way we would still have the same flexibility, without misusing modopts. Of course that would be a structural file format breaking change, but not bigger than changing the meaning of the callconv byte to be a compressed integer. Either way, that ship has sailed.

Reply:
I brought this up since it was discussed in the 7.0 PR and thought continuing that discussion may be worthwhile.

Also FWIW I came across [these comments](https://github.com/dotnet/runtime/blob/4cd3e780c469e2416b6ebaba903dbea7c411a5b8/src/coreclr/vm/siginfo.cpp#L3742-L3746) that mentioned we used to read the calling conventions as a compressed integer.